src/main/java/com/algorand/algosdk/util/abi/types/Type.java

```diff
@@ -0,0 +1,133 @@
+            for (String subStr : tupleContent)
+                tupleTypes.add(Type.fromString(subStr));
+            return new TypeTuple(tupleTypes);
+        } else {
+            throw new IllegalArgumentException("Cannot infer type from the string: " + str);
+        }
+    }
+
+    public static class Segment {
+        public int L, R;
+
+        Segment(int left, int right) {
+            this.L = left;
+            this.R = right;
+        }
+    }
+
+    private static List<String> parseTupleContent(String str) {
+        if (str.length() == 0)
+            return new ArrayList<>();
+
+        if (str.startsWith(",") || str.endsWith(","))
+            throw new IllegalArgumentException("parsing error: tuple content should not start with comma");
+
+        if (str.contains(",,"))
+            throw new IllegalArgumentException("parsing error: tuple content should not have consecutive commas");
+
+        Stack<Integer> parenStack = new Stack<>();
+        List<Segment> parenSegments = new ArrayList<>();
+
+        for (int i = 0; i < str.length(); i++) {
+            if (str.charAt(i) == '(')
+                parenStack.push(i);
+            else if (str.charAt(i) == ')') {
+                if (parenStack.empty())
+                    throw new IllegalArgumentException("parsing error: tuple parentheses are not balanced: " + str);
+                int leftParenIndex = parenStack.pop();
+                if (parenStack.empty())
+                    parenSegments.add(new Segment(leftParenIndex, i));
+            }
+        }
+        if (!parenStack.empty())
+            throw new IllegalArgumentException("parsing error: tuple parentheses are not balanced: " + str);
+
+        String strCopied = str;
+        for (int i = parenSegments.size() - 1; i >= 0; i--)
+            strCopied = strCopied.substring(0, parenSegments.get(i).L) + strCopied.substring(parenSegments.get(i).R + 1);
+
+        String[] tupleSeg = strCopied.split(",", -1);
+        int parenSegCount = 0;
+        for (int i = 0; i < tupleSeg.length; i++) {
+            if (tupleSeg[i].isEmpty()) {
+                tupleSeg[i] = str.substring(parenSegments.get(parenSegCount).L, parenSegments.get(parenSegCount).R + 1);
+                parenSegCount++;
+            }
+        }
+
+        return Arrays.asList(tupleSeg);
+    }
+
+    public static int findBoolLR(Type[] typeArray, int index, int delta) {
+        int until = 0;
+        while (true) {
+            int currentIndex = index + delta * until;
+            if (typeArray[currentIndex] instanceof TypeBool) {
+                if (currentIndex != typeArray.length - 1 && delta > 0)
+                    until++;
+                else if (currentIndex != 0 && delta < 0)
+                    until++;
+                else
+                    break;
+            } else {
+                until--;
+                break;
+            }
+        }
+        return until;
+    }
+}
```

Comment:
nit: Is there a linter/formatter for the Java SDK? Github is showing that some of the files need a newline at the end.

Reply:
have to manually add the lines, I doubt if there's code formatter/linter here in JAVA SDK...